libs/qCC_io/ShpFilter.cpp

```diff
@@ -1,1846 +1,1892 @@
-									if (poly)
-									{
-										poly->applyGLTransformation_recursive(&trans);
-										//this transformation is of no interest for the user
-										poly->resetGLTransformationHistory_recursive();
-										//add the 'const altitude' meta-data as well
-										poly->setMetaData(ccPolyline::MetaKeyConstAltitude(), QVariant(z));
-									}
-								}
-							}
-							else if (hasPoints)
-							{
-								//for each point
-								for (unsigned i = 0; i < singlePoints->size(); ++i)
-								{
-									//get the height
-									double z = 0.0;
-									if (fieldType == FTDouble)
-										z = DBFReadDoubleAttribute(dbfHandle, static_cast<int>(i), index);
-									else //if (fieldType == FTInteger)
-										z = static_cast<double>(DBFReadIntegerAttribute(dbfHandle, static_cast<int>(i), index));
-									z *= scale;
-
-									//set the point height
-									const_cast<CCVector3*>(singlePoints->getPoint(i))->z = z;
-								}
-								singlePoints->invalidateBoundingBox();
-							}
-							else
-							{
-								assert(false);
-							}
-						}
-					}
-				}
-				else
-				{
-					ccLog::Warning("[SHP] No numerical field in the associated DBF file!");
-				}
-			}
-		}
-		else
-		{
-			ccLog::Warning(QString("[SHP] Failed to load associated DBF file ('%1')").arg(dbfFilename));
-		}
-	}
-
-	if (singlePoints)
-	{
-		if (singlePoints->size() == 0)
-		{
-			delete singlePoints;
-			singlePoints = 0;
-		}
-		else
-		{
-			CCLib::ScalarField* sf = singlePoints->getScalarField(0);
-			if (sf)
-			{
-				sf->computeMinAndMax();
-				singlePoints->showSF(true);
-			}
-			container.addChild(singlePoints);
-		}
-	}
-
-	return error;
-}
-
-#endif //CC_SHP_SUPPORT
+//##########################################################################
+//#                                                                        #
+//#                              CLOUDCOMPARE                              #
+//#                                                                        #
+//#  This program is free software; you can redistribute it and/or modify  #
+//#  it under the terms of the GNU General Public License as published by  #
+//#  the Free Software Foundation; version 2 or later of the License.      #
+//#                                                                        #
+//#  This program is distributed in the hope that it will be useful,       #
+//#  but WITHOUT ANY WARRANTY; without even the implied warranty of        #
+//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the          #
+//#  GNU General Public License for more details.                          #
+//#                                                                        #
+//#          COPYRIGHT: EDF R&D / TELECOM ParisTech (ENST-TSI)             #
+//#                                                                        #
+//##########################################################################
+
+#ifdef CC_SHP_SUPPORT

```

Comment:
Are you using space-based tabulations maybe? It's hard to see the differences 😢

Reply:
I use tabs for indentation, so it should be tabs 🤔

Reply:
Do you have a clue why Git(hub) believes everything has changed then?

Reply:
It may be because I changed a lot of stuff.
I tried to do a pull request on my own [fork,](https://github.com/tmontaigu/CloudCompare/pull/1/files) the diff does not look as if everything changed, but is very hard to follow

Reply:
I think it's just the number of changes in that file.

FWIW this is just too large for me to review. In the future, a series of smaller, logical, atomic PRs would be much more helpful (i.e. one might be "move enums into header" - that size). Testing stuff could be a completely separate series of PRs.

I can't speak for the others, but for me something that takes 5 minutes is much more likely to happen than something that's going to take me 1+ hours.

I do appreciate the contribution! Just want to give feedback since I believe we need more code reviews.

Reply:
Then, maybe I could try do redo the changes in multiple PR's

Reply:
If you can, I think it would help in terms of understanding what's in this big change.

Maybe start with the "cleanup/style" type things you did in preparation for the guts of your change if you can? These are usually easy to review.

I would start by separating the testing code out to do after, then look at the larger change for the smaller logical pieces - might be "Reformat case statements for readability", "Move enums into header and change into class-based enum", "Move member initialization into header",  or "Document undocumented methods".

If there's anything I can help with, please let me know.

Reply:
So I should make the following PRs:
1) General code modernization (enum class, nullptr, etc)
2) Modernization of the file reading (Use QDataStream, split reading functions, fix the 2 bugs I had when reading .shp files)
2) Modernization of the file writing (same idea as above but will introduce more changes)
3) Add Reading & Writing for Multipatch shapes
4) Add tests

Reply:
I think that sounds great.

When I'm reviewing code, I like to be able to focus on the one thing that it's trying to accomplish. Separating it like this would help a lot.

Thanks!